### DISCUSSION TITLE
next-pwa vs next-offline

### DISCUSSION THREAD

**Discussion by Ojay** [vercel/next.js#13810] [Help] [answered]
Hi all,

We're looking to setup our app as a PWA, mainly because we need some offline capabilities. I know this is a little vague of a question and probably comes down to project specifics, but, I wondered if anyone had any insight on how [next-pwa](https://www.npmjs.com/package/next-pwa) and [next-offline](https://www.npmjs.com/package/next-offline) stack up against each other?

Is there another best practice alternative? Just wondering if there's a bit of a consensus surrounding serving content offline with Next.

Thanks so much for your help 🤙

**Comment by Ojay** [comment]
Cheeky bump 😬

**Comment by jensmeindertsma** [comment]
I went with `next-offline` because `next-pwa` also enables the service worker during development, which was getting really annoying for me. That also makes it very slow in development.

**Reply by Ojay** [reply]
Thanks for the feedback, much appreciated.

**Reply by gramsco** [reply]
next-pwa use the example below as an example of a config file [here](https://www.npmjs.com/package/next-pwa#configuration)

```javascript
const withPWA = require('next-pwa')
 
module.exports = withPWA({
  pwa: {
    disable: process.env.NODE_ENV === 'development',
    register: true,
    scope: '/app',
    sw: 'service-worker.js',
    //...
  }
})
```
Isn't that enough to prevent this from happening?

**Comment by ankesh7** [comment]
We are using next-offline and it is working well with a custom config. Haven't used next-pwa although but the runtime strategy looks very promising that is ideal for most web applications like NetworkFirst for /api/ and staleWhileRevalidate strategy for fonts and images, etc. 

Next-offline in general implements caching for every resource with `https://` and high limit on entries which made us fall under trap with video segments from external origins being cached leading to storage quota errors for many of the users.

I would say at the end you should always have your own strategies and patterns to cache resources which next-pwa is great at doing. If I had to use next-pwa, I would still go about having a strategy that only caches resources from my origin because it fits well to my personal use-case and caching everything is not ideal.

**Comment by redbar0n** [comment]
next-offline vs next-pwa:

next-pwa _“is more opinionated (with presets of workbox configurations) and aims to be as simple as possible for users.”_ [according to the author himself](https://github.com/shadowwalker/next-pwa/issues/14).

next-offline is "more mature", based on having existed 2 years longer (since 2018), according to npmtrends.com.

next-offline is more popular, but both are rising, and next-pwa is rising faster:
* npm trends: [https://www.npmtrends.com/next-offline-vs-next-pwa](https://www.npmtrends.com/next-offline-vs-next-pwa)
* star history: [https://star-history.t9t.io/#hanford/next-offline&shadowwalker/next-pwa](https://star-history.t9t.io/#hanford/next-offline&shadowwalker/next-pwa)

Bundle sizes (min & gzip) for both are quite large (potentially prohibitively so, imho):
* next-pwa 5.0.0 takes [787.3 kB](https://bundlephobia.com/result?p=next-pwa@5.0.0). Update: next-pwa 5.4.0 increased to [1.4 MB](https://bundlephobia.com/package/next-pwa@5.4.0)! See: [discussion](https://github.com/shadowwalker/next-pwa/discussions/290).
* next-offline 5.0.3 takes [916.3 kB](https://bundlephobia.com/result?p=next-offline@5.0.3). Update: next-offline 5.0.5 increased to [1.4 MB](https://bundlephobia.com/package/next-offline@5.0.5)!

If you are using Expo Web, then next-offline is the one they refer to in the [expo web docs](https://docs.expo.io/guides/using-nextjs/), so it seems a safer bet in that case.

**Reply by redbar0n** [reply]
Now, over 6 months later, it seems that _next-offline_ has significantly waned in popularity, and download numbers are shrinking. While _next-pwa_ is continuing its meteoric rise in downloads:

https://www.npmtrends.com/next-offline-vs-next-pwa

For whatever that is worth.

**Comment by milad-afkhami** [comment]
Beside all the answers, i prefer using **next-pwa** for this particular reason:

- Unlike **next-offline** it will not force us to write a custom server, which we now from [next.js docs](https://nextjs.org/docs/advanced-features/custom-server) that :

> A custom server will remove important performance optimizations, like serverless functions and Automatic Static Optimization.

**Reply by lubojanski** [reply]
but does next-offline force you to write custom server?

**Reply by milad-afkhami** [reply]
Yeah i think so
based on [next-offline docs](https://github.com/hanford/next-offline#now-10) 

> You'll want to use the next.js custom server API. The easiest way to do that is creating a server.js that looks like this:

Am i right or is it just for serving **service-worker.js** and nothing more?

**Comment by zacBkh** [comment]
Hello everyone,

I am trying to create a PWA using next js with offline support **without any libraries**
I don't like the abstraction they provide and all our app is becoming instantly dependant on someone's else code. 
By the way I believe next offline and so many others are not maintained anymore. 

My PWA is installable and works well however when I want to cache resources, I have an issue. 
In a traditional website without framework, we would cache any page by specifying it's relative path, for example : ["/offline-page.html"] and we would add this to cache.All or something like that. 
But in next js, how can we do this since our pages are served through many cryptic files bundled in chunks?

My app is working well offline if I put a "offline.html" file in my /public folder.
But then we loose the benefit of next and react because we cannot use jsx, local states etc etc. 

Am I missing something here?

**Reply by thijssmudde** [reply]
@zacBkh Did you find any solution for making your Nextjs app work offline?

**Comment by leerob** [comment, marked as answer]
Hey everyone. We now have updated documentation on building PWAs with Next.js.

https://nextjs.org/docs/app/building-your-application/configuring/progressive-web-apps

**Reply by Yevhenns** [reply]
Hey. What's the problem? I'm getting an error in actions.ts - await webpush.sendNotification(subscription:
Argument of type 'PushSubscription' is not assignable to parameter of type 'import(".../node_modules/@types/web-push/index").PushSubscription'.
  Property 'keys' is missing in type 'PushSubscription' but required in type 'import(".../node_modules/@types/web-push/index").PushSubscription'.ts(2345)
index.d.ts(209, 5): 'keys' is declared here.

**Reply by kennethhaspeel** [reply]
I'm having the same problem as @Yevhenns is having. The code works, the notification is sent in development. But on deployment to Vercel the build fails.

When i console log subscription in the sendNotification function, the keys are present. So not really sure what the problem could be

**Reply by kennethhaspeel** [reply]
i think i found the solution. You also have to import the correct type
`import  webPush, {PushSubscription} from 'web-push'`

If you do not import this PushSubscription from web-push, it uses a default one from typescript (i think) which doesn't have the keys property.

I tried editing the doc page but not sure if i did everything correct

**Reply by gcb** [reply]
How can this be the accepted answer if it doesn't even answer the question here which was "next-pwa vs next-offline" :) the page lists everything from PWA *but* offline support

the only mention of offline support in a footer note and mention serwist, so should we assume only @serwist/next is "officially supported"?

**Reply by dantman** [reply]
Yes I also find the lack of detail on offline support terrible for something that was supposed to be an answer to this question. Worse it seems to be outdated. Here's what I found after digging myself:

1. [next-pwa](https://www.npmjs.com/package/next-pwa) has not been updated in 3 years and [next-offline](https://www.npmjs.com/package/next-offline) has not been updated in 5 years, so I guess they are correct to not list them as options
2. Both next-pwa and next-offline seem to be Next.js specific wrappers around Google's [Workbox](https://developer.chrome.com/docs/workbox/) library. Which Serwist's GitHub says Serwist "It is a fork of Workbox that came to be due to the its development being stagnated. Most of the prior work was done by Google LLC's team and the community."
3. Next.js documentation states "Note: this plugin currently requires webpack configuration." however counter to that [@serwist/next's Getting Started page](https://serwist.pages.dev/docs/next/getting-started) points to a [TurboPack Getting Started page](https://serwist.pages.dev/docs/next/turbo) suggesting this statement is false and may be driving people (like me) to go research next-pwa/next-offline as other options when Serwist might actually be suitable.

**Reply by gcb** [reply]
Thanks for sharing. Just keep in mind turbopack is not ready for standalone deployments of nextjs, which you likely are doing if you are here.
